In [1]:
from google.colab import files

uploaded = files.upload()

Saving Day10_Flight_Operations_Dataset.csv to Day10_Flight_Operations_Dataset (1).csv


In [2]:
import pandas as pd

df = pd.read_csv("Day10_Flight_Operations_Dataset.csv")

df.head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5


In [3]:
print("First 5 rows:")
display(df.head())

print("\nLast 5 rows:")
display(df.tail())

First 5 rows:


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5



Last 5 rows:


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
175,FL0176,2026-04-10,IndiGo,Ahmedabad,Bengaluru,Airbus A320,Economy,48,189,4298,0.0,On Time,Storm,Mobile App,11,Vegetarian,5
176,FL0177,2026-02-17,Vistara,Chennai,Mumbai,Airbus A320,Business,124,180,17306,20.0,Delayed,Cloudy,Mobile App,8,No Meal,5
177,FL0178,2026-04-30,IndiGo,Delhi,Mumbai,Airbus A319,Economy,112,186,4490,110.0,Delayed,Fog,Online Travel Portal,16,Vegetarian,3
178,FL0179,2026-05-08,Vistara,Bengaluru,Kochi,Airbus A320,Economy,134,189,5413,20.0,Delayed,Rain,Online Travel Portal,9,No Meal,2
179,FL0180,2026-01-22,Akasa Air,Delhi,Mumbai,Airbus A321,Economy,144,210,3767,75.0,Delayed,Fog,Online Travel Portal,17,No Meal,5


In [4]:
print("Dataset Shape:")
print(df.shape)

print("\nColumn Names:")
print(df.columns.tolist())

print("\nData Types:")
print(df.dtypes)

Dataset Shape:
(180, 17)

Column Names:
['Flight_ID', 'Flight_Date', 'Airline', 'Origin', 'Destination', 'Aircraft', 'Travel_Class', 'Passengers', 'Seat_Capacity', 'Average_Ticket_Price', 'Delay_Minutes', 'Flight_Status', 'Weather', 'Booking_Channel', 'Avg_Baggage_Kg', 'Meal_Preference', 'Passenger_Satisfaction']

Data Types:
Flight_ID                  object
Flight_Date                object
Airline                    object
Origin                     object
Destination                object
Aircraft                   object
Travel_Class               object
Passengers                  int64
Seat_Capacity               int64
Average_Ticket_Price        int64
Delay_Minutes             float64
Flight_Status              object
Weather                    object
Booking_Channel            object
Avg_Baggage_Kg              int64
Meal_Preference            object
Passenger_Satisfaction      int64
dtype: object


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Flight_ID               180 non-null    object 
 1   Flight_Date             180 non-null    object 
 2   Airline                 180 non-null    object 
 3   Origin                  180 non-null    object 
 4   Destination             180 non-null    object 
 5   Aircraft                180 non-null    object 
 6   Travel_Class            180 non-null    object 
 7   Passengers              180 non-null    int64  
 8   Seat_Capacity           180 non-null    int64  
 9   Average_Ticket_Price    180 non-null    int64  
 10  Delay_Minutes           173 non-null    float64
 11  Flight_Status           180 non-null    object 
 12  Weather                 180 non-null    object 
 13  Booking_Channel         180 non-null    object 
 14  Avg_Baggage_Kg          180 non-null    in

In [6]:
print("Missing Values:")
print(df.isnull().sum())

print("\nDuplicate Rows:")
print(df.duplicated().sum())

Missing Values:
Flight_ID                 0
Flight_Date               0
Airline                   0
Origin                    0
Destination               0
Aircraft                  0
Travel_Class              0
Passengers                0
Seat_Capacity             0
Average_Ticket_Price      0
Delay_Minutes             7
Flight_Status             0
Weather                   0
Booking_Channel           0
Avg_Baggage_Kg            0
Meal_Preference           0
Passenger_Satisfaction    0
dtype: int64

Duplicate Rows:
0


In [7]:
print("Missing Value Percentage:")
print((df.isnull().sum() / len(df) * 100).round(2))

Missing Value Percentage:
Flight_ID                 0.00
Flight_Date               0.00
Airline                   0.00
Origin                    0.00
Destination               0.00
Aircraft                  0.00
Travel_Class              0.00
Passengers                0.00
Seat_Capacity             0.00
Average_Ticket_Price      0.00
Delay_Minutes             3.89
Flight_Status             0.00
Weather                   0.00
Booking_Channel           0.00
Avg_Baggage_Kg            0.00
Meal_Preference           0.00
Passenger_Satisfaction    0.00
dtype: float64


In [8]:
print("Statistical Summary:")
df.describe()

Statistical Summary:


,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Avg_Baggage_Kg,Passenger_Satisfaction
count,180.000000,180.000000,180.000000,173.000000,180.000000,180.000000
mean,122.772222,197.238889,6111.677778,19.335260,15.150000,3.494444
std,45.635651,15.408176,3615.220040,29.029712,5.705506,1.169908
min,45.000000,180.000000,1949.000000,0.000000,5.000000,2.000000
25%,78.750000,186.000000,3815.500000,0.000000,10.000000,2.000000
50%,125.000000,189.000000,5380.000000,5.000000,15.500000,3.000000
75%,162.250000,210.000000,6790.000000,20.000000,20.000000,5.000000
max,210.000000,220.000000,27832.000000,110.000000,25.000000,5.000000


In [9]:
print("Categorical Column Summary:")
df.describe(include='object')

Categorical Column Summary:


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Flight_Status,Weather,Booking_Channel,Meal_Preference
count,180,180,180,180,180,180,180,180,180,180,180
unique,180,111,6,11,8,4,3,3,5,4,4
top,FL0001,2026-02-10,Vistara,Delhi,Delhi,Airbus A319,Economy,On Time,Rain,Online Travel Portal,Vegetarian
freq,1,6,37,43,51,60,134,114,43,50,52


In [12]:
print("Single Column:")
display(df['Flight_Status'])

print("Multiple Columns:")
display(df[['Flight_Status', 'Airline']])

print("First 5 Rows:")
display(df.iloc[0:5])

print("First 5 Rows and First 3 Columns:")
display(df.iloc[0:5, 0:3])

print("First Row:")
display(df.loc[0])

Single Column:


,Flight_Status
0,Delayed
1,On Time
2,On Time
3,Delayed
4,Delayed
...,...
175,On Time
176,Delayed
177,Delayed
178,Delayed


Multiple Columns:


,Flight_Status,Airline
0,Delayed,SpiceJet
1,On Time,Akasa Air
2,On Time,Air India
3,Delayed,Akasa Air
4,Delayed,Air India
...,...,...
175,On Time,IndiGo
176,Delayed,Vistara
177,Delayed,IndiGo
178,Delayed,Vistara


First 5 Rows:


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5


First 5 Rows and First 3 Columns:


,Flight_ID,Flight_Date,Airline
0,FL0001,2026-03-23,SpiceJet
1,FL0002,2026-06-07,Akasa Air
2,FL0003,2026-05-12,Air India
3,FL0004,2026-05-30,Akasa Air
4,FL0005,2026-02-27,Air India


First Row:


,0
Flight_ID,FL0001
Flight_Date,2026-03-23
Airline,SpiceJet
Origin,Mumbai
Destination,Bengaluru
Aircraft,Airbus A319
Travel_Class,Economy
Passengers,210
Seat_Capacity,180
Average_Ticket_Price,6894


In [13]:
print("On Time Flights:")
display(df[df['Flight_Status'] == 'On Time'])

print("Delayed Flights:")
display(df[df['Flight_Status'] == 'Delayed'])

print("Cancelled Flights:")
display(df[df['Flight_Status'] == 'Cancelled'])

print("Flight Status Counts:")
display(df['Flight_Status'].value_counts())

On Time Flights:


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2
6,FL0007,2026-06-17,Air India Express,Delhi,Srinagar,Airbus A319,Premium Economy,178,210,9430,0.0,On Time,Storm,Online Travel Portal,18,Vegetarian,4
7,FL0008,2026-04-28,Air India,Delhi,Chennai,Airbus A319,Business,157,220,14173,0.0,On Time,Fog,Airline Website,25,Vegan,5
9,FL0010,2026-06-12,Akasa Air,Ahmedabad,Bengaluru,Airbus A321,Economy,125,220,2500,12.0,On Time,Cloudy,Online Travel Portal,23,Non-Vegetarian,4
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
166,FL0167,2026-06-15,SpiceJet,Ahmedabad,Bengaluru,Airbus A321,Economy,148,180,6593,0.0,On Time,Rain,Online Travel Portal,15,Non-Vegetarian,5
167,FL0168,2026-05-01,Air India,Delhi,Mumbai,Airbus A319,Economy,47,220,6479,0.0,On Time,Cloudy,Mobile App,18,Vegan,4
171,FL0172,2026-06-02,IndiGo,Pune,Delhi,Airbus A319,Premium Economy,166,180,9210,12.0,On Time,Fog,Airline Website,17,Vegetarian,2
174,FL0175,2026-04-23,Akasa Air,Bengaluru,Kochi,Airbus A321,Economy,116,180,5883,0.0,On Time,Cloudy,Online Travel Portal,14,Non-Vegetarian,5


Delayed Flights:


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5
5,FL0006,2026-05-20,Akasa Air,Kochi,Delhi,Airbus A319,Economy,123,220,5636,35.0,Delayed,Cloudy,Travel Agency,13,Vegetarian,2
8,FL0009,2026-01-25,Akasa Air,Delhi,Chennai,Airbus A320,Economy,82,210,5893,75.0,Delayed,Cloudy,Mobile App,8,Vegetarian,4
16,FL0017,2026-05-24,SpiceJet,Delhi,Chennai,Airbus A321,Economy,73,220,4256,20.0,Delayed,Rain,Travel Agency,22,Non-Vegetarian,2
17,FL0018,2026-03-09,Air India Express,Jaipur,Mumbai,Airbus A321,Economy,88,180,6133,20.0,Delayed,Clear,Travel Agency,13,Vegetarian,4
21,FL0022,2026-02-01,Vistara,Kochi,Delhi,Airbus A321,Economy,126,220,5887,75.0,Delayed,Clear,Airline Website,10,Vegetarian,2
26,FL0027,2026-02-10,SpiceJet,Bengaluru,Hyderabad,Airbus A319,Economy,169,180,3973,110.0,Delayed,Clear,Mobile App,7,Vegan,4
29,FL0030,2026-04-10,SpiceJet,Delhi,Srinagar,Airbus A321,Economy,200,210,2716,75.0,Delayed,Storm,Airline Website,10,Vegan,3


Cancelled Flights:


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
23,FL0024,2026-05-19,SpiceJet,Ahmedabad,Bengaluru,Airbus A319,Economy,110,186,5578,NaN,Cancelled,Rain,Travel Agency,9,Vegetarian,3
37,FL0038,2026-03-27,Air India,Delhi,Srinagar,Airbus A320,Economy,166,210,5972,NaN,Cancelled,Rain,Travel Agency,5,Non-Vegetarian,5
40,FL0041,2026-03-23,Akasa Air,Delhi,Chennai,Airbus A321,Economy,170,189,4013,NaN,Cancelled,Fog,Travel Agency,13,No Meal,3
65,FL0066,2026-03-06,SpiceJet,Srinagar,Delhi,Airbus A320,Economy,83,210,4418,NaN,Cancelled,Cloudy,Travel Agency,19,Vegetarian,5
68,FL0069,2026-03-09,SpiceJet,Kochi,Delhi,Airbus A321,Economy,107,186,5911,NaN,Cancelled,Cloudy,Online Travel Portal,23,Non-Vegetarian,5
110,FL0111,2026-05-16,Vistara,Delhi,Mumbai,Airbus A321,Premium Economy,124,186,6711,NaN,Cancelled,Cloudy,Mobile App,19,Vegetarian,4
160,FL0161,2026-04-03,Vistara,Delhi,Mumbai,Airbus A320,Economy,135,220,6154,NaN,Cancelled,Storm,Mobile App,6,Vegetarian,2


Flight Status Counts:


,count
Flight_Status,
On Time,114
Delayed,59
Cancelled,7


In [14]:
print("Sorted by Airline:")
display(df.sort_values(by='Airline'))

print("Sorted by Airline (Descending):")
display(df.sort_values(by='Airline', ascending=False))

print("Sorted by Flight Status:")
display(df.sort_values(by='Flight_Status'))

print("Sorted by Airline and Flight Status:")
display(df.sort_values(by=['Airline', 'Flight_Status']))

Sorted by Airline:


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
121,FL0122,2026-02-13,Air India,Mumbai,Kolkata,Airbus A319,Economy,178,189,5514,0.0,On Time,Rain,Airline Website,25,Non-Vegetarian,5
98,FL0099,2026-01-15,Air India,Mumbai,Bengaluru,Boeing 737,Economy,108,189,3588,35.0,Delayed,Rain,Airline Website,14,Non-Vegetarian,4
25,FL0026,2026-05-11,Air India,Ahmedabad,Bengaluru,Boeing 737,Premium Economy,137,186,8836,12.0,On Time,Fog,Online Travel Portal,22,No Meal,4
167,FL0168,2026-05-01,Air India,Delhi,Mumbai,Airbus A319,Economy,47,220,6479,0.0,On Time,Cloudy,Mobile App,18,Vegan,4
164,FL0165,2026-03-31,Air India,Ahmedabad,Bengaluru,Airbus A320,Economy,137,220,4992,12.0,On Time,Clear,Mobile App,18,Vegetarian,3
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
66,FL0067,2026-04-23,Vistara,Mumbai,Kolkata,Boeing 737,Premium Economy,109,210,8473,20.0,Delayed,Cloudy,Airline Website,24,Vegetarian,3
63,FL0064,2026-06-24,Vistara,Mumbai,Bengaluru,Airbus A321,Economy,105,180,4280,0.0,On Time,Storm,Airline Website,5,Vegan,3
149,FL0150,2026-06-15,Vistara,Chennai,Mumbai,Boeing 737,Economy,131,210,6572,5.0,On Time,Storm,Mobile App,25,No Meal,5
42,FL0043,2026-06-19,Vistara,Delhi,Chennai,Airbus A319,Economy,64,220,4029,50.0,Delayed,Clear,Mobile App,5,No Meal,3


Sorted by Airline (Descending):


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
90,FL0091,2026-02-21,Vistara,Bengaluru,Kochi,Airbus A320,Economy,52,189,2251,0.0,On Time,Fog,Online Travel Portal,15,Non-Vegetarian,2
39,FL0040,2026-03-21,Vistara,Kochi,Delhi,Airbus A319,Economy,178,220,5361,12.0,On Time,Cloudy,Mobile App,20,Vegan,4
63,FL0064,2026-06-24,Vistara,Mumbai,Bengaluru,Airbus A321,Economy,105,180,4280,0.0,On Time,Storm,Airline Website,5,Vegan,3
30,FL0031,2026-01-08,Vistara,Delhi,Srinagar,Airbus A321,Premium Economy,209,180,9962,50.0,Delayed,Cloudy,Mobile App,9,Vegetarian,4
31,FL0032,2026-01-28,Vistara,Hyderabad,Kolkata,Airbus A319,Premium Economy,142,189,6730,12.0,On Time,Fog,Airline Website,22,Non-Vegetarian,4
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
101,FL0102,2026-01-05,Air India,Jaipur,Mumbai,Airbus A319,Economy,158,180,2991,0.0,On Time,Cloudy,Airline Website,16,No Meal,2
131,FL0132,2026-02-04,Air India,Hyderabad,Kolkata,Airbus A321,Economy,61,220,3483,75.0,Delayed,Clear,Mobile App,18,No Meal,5
25,FL0026,2026-05-11,Air India,Ahmedabad,Bengaluru,Boeing 737,Premium Economy,137,186,8836,12.0,On Time,Fog,Online Travel Portal,22,No Meal,4
125,FL0126,2026-03-04,Air India,Bengaluru,Kochi,Airbus A319,Economy,125,210,2516,5.0,On Time,Clear,Online Travel Portal,15,No Meal,5


Sorted by Flight Status:


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
110,FL0111,2026-05-16,Vistara,Delhi,Mumbai,Airbus A321,Premium Economy,124,186,6711,NaN,Cancelled,Cloudy,Mobile App,19,Vegetarian,4
65,FL0066,2026-03-06,SpiceJet,Srinagar,Delhi,Airbus A320,Economy,83,210,4418,NaN,Cancelled,Cloudy,Travel Agency,19,Vegetarian,5
40,FL0041,2026-03-23,Akasa Air,Delhi,Chennai,Airbus A321,Economy,170,189,4013,NaN,Cancelled,Fog,Travel Agency,13,No Meal,3
68,FL0069,2026-03-09,SpiceJet,Kochi,Delhi,Airbus A321,Economy,107,186,5911,NaN,Cancelled,Cloudy,Online Travel Portal,23,Non-Vegetarian,5
160,FL0161,2026-04-03,Vistara,Delhi,Mumbai,Airbus A320,Economy,135,220,6154,NaN,Cancelled,Storm,Mobile App,6,Vegetarian,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
88,FL0089,2026-03-15,Air India Express,Pune,Delhi,Boeing 737,Premium Economy,57,220,10062,0.0,On Time,Clear,Travel Agency,9,Vegan,5
87,FL0088,2026-02-08,Vistara,Delhi,Srinagar,Airbus A321,Economy,75,210,4156,0.0,On Time,Cloudy,Travel Agency,20,Vegetarian,3
86,FL0087,2026-02-26,SpiceJet,Chennai,Mumbai,Airbus A320,Economy,68,189,4704,5.0,On Time,Cloudy,Online Travel Portal,22,Vegan,3
118,FL0119,2026-02-10,Air India Express,Delhi,Chennai,Boeing 737,Economy,66,189,3705,5.0,On Time,Clear,Mobile App,14,Vegetarian,2


Sorted by Airline and Flight Status:


,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
37,FL0038,2026-03-27,Air India,Delhi,Srinagar,Airbus A320,Economy,166,210,5972,NaN,Cancelled,Rain,Travel Agency,5,Non-Vegetarian,5
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5
47,FL0048,2026-01-08,Air India,Delhi,Chennai,Airbus A321,Economy,68,210,3697,20.0,Delayed,Fog,Mobile App,16,Non-Vegetarian,5
50,FL0051,2026-05-03,Air India,Delhi,Mumbai,Airbus A321,Economy,135,210,5129,110.0,Delayed,Rain,Online Travel Portal,17,Vegetarian,4
84,FL0085,2026-05-03,Air India,Lucknow,Delhi,Airbus A319,Economy,45,189,3299,35.0,Delayed,Rain,Online Travel Portal,21,Non-Vegetarian,3
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
91,FL0092,2026-06-23,Vistara,Delhi,Chennai,Airbus A321,Economy,170,210,6486,0.0,On Time,Rain,Travel Agency,17,Vegetarian,2
94,FL0095,2026-06-14,Vistara,Bengaluru,Kochi,Airbus A319,Economy,199,189,4369,0.0,On Time,Fog,Online Travel Portal,11,Vegetarian,5
123,FL0124,2026-01-06,Vistara,Delhi,Chennai,Airbus A319,Economy,138,210,5399,0.0,On Time,Cloudy,Mobile App,11,No Meal,3
149,FL0150,2026-06-15,Vistara,Chennai,Mumbai,Boeing 737,Economy,131,210,6572,5.0,On Time,Storm,Mobile App,25,No Meal,5


In [15]:
print("Number of Flights by Airline:")
display(df.groupby('Airline').size())

print("Flights by Airline and Status:")
display(df.groupby(['Airline', 'Flight_Status']).size())

print("Flight Status Distribution by Airline:")
display(df.groupby('Airline')['Flight_Status'].value_counts())

print("Number of Flights by Status:")
display(df.groupby('Flight_Status').size())

Number of Flights by Airline:


,0
Airline,
Air India,24
Air India Express,24
Akasa Air,31
IndiGo,32
SpiceJet,32
Vistara,37


Flights by Airline and Status:


Airline            Flight_Status
Air India          Cancelled         1
                   Delayed           9
                   On Time          14
Air India Express  Delayed           6
                   On Time          18
Akasa Air          Cancelled         1
                   Delayed           8
                   On Time          22
IndiGo             Delayed          11
                   On Time          21
SpiceJet           Cancelled         3
                   Delayed          12
                   On Time          17
Vistara            Cancelled         2
                   Delayed          13
                   On Time          22
dtype: int64

Flight Status Distribution by Airline:


Airline            Flight_Status
Air India          On Time          14
                   Delayed           9
                   Cancelled         1
Air India Express  On Time          18
                   Delayed           6
Akasa Air          On Time          22
                   Delayed           8
                   Cancelled         1
IndiGo             On Time          21
                   Delayed          11
SpiceJet           On Time          17
                   Delayed          12
                   Cancelled         3
Vistara            On Time          22
                   Delayed          13
                   Cancelled         2
Name: count, dtype: int64

Number of Flights by Status:


,0
Flight_Status,
Cancelled,7
Delayed,59
On Time,114


In [16]:
status_mapping = {
    'On Time': 1,
    'Delayed': 2,
    'Cancelled': 3
}

df['Status_Code'] = df['Flight_Status'].map(status_mapping)

df[['Flight_Status', 'Status_Code']].head()

,Flight_Status,Status_Code
0,Delayed,2
1,On Time,1
2,On Time,1
3,Delayed,2
4,Delayed,2


In [17]:
df['Airline_Status'] = df['Airline'] + ' - ' + df['Flight_Status']

df[['Airline', 'Flight_Status', 'Airline_Status']].head()

,Airline,Flight_Status,Airline_Status
0,SpiceJet,Delayed,SpiceJet - Delayed
1,Akasa Air,On Time,Akasa Air - On Time
2,Air India,On Time,Air India - On Time
3,Akasa Air,Delayed,Akasa Air - Delayed
4,Air India,Delayed,Air India - Delayed


In [18]:
df_transformed = df.copy()

df_transformed['Status_Code'] = df_transformed['Flight_Status'].map(status_mapping)

df_transformed.head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Status_Code,Airline_Status
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3,2,SpiceJet - Delayed
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5,1,Akasa Air - On Time
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2,1,Air India - On Time
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2,2,Akasa Air - Delayed
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5,2,Air India - Delayed


In [19]:
airline_counts = df['Airline'].value_counts()

print("Flights by Airline:")
display(airline_counts)

print("\nAirline with the Most Flights:")
print(airline_counts.idxmax())

Flights by Airline:


,count
Airline,
Vistara,37
SpiceJet,32
IndiGo,32
Akasa Air,31
Air India Express,24
Air India,24



Airline with the Most Flights:
Vistara


In [20]:
status_counts = df['Flight_Status'].value_counts()

print("Flight Status Distribution:")
display(status_counts)

print("\nMost Common Flight Status:")
print(status_counts.idxmax())

Flight Status Distribution:


,count
Flight_Status,
On Time,114
Delayed,59
Cancelled,7



Most Common Flight Status:
On Time


In [22]:
delayed_flights = df[df['Flight_Status'] == 'Delayed']

print("Delayed Flights by Airline:")
display(delayed_flights['Airline'].value_counts())

Delayed Flights by Airline:


,count
Airline,
Vistara,13
SpiceJet,12
IndiGo,11
Air India,9
Akasa Air,8
Air India Express,6


In [23]:
cancelled_flights = df[df['Flight_Status'] == 'Cancelled']

print("Cancelled Flights by Airline:")
display(cancelled_flights['Airline'].value_counts())

Cancelled Flights by Airline:


,count
Airline,
SpiceJet,3
Vistara,2
Air India,1
Akasa Air,1


In [24]:
delayed_counts = df[df['Flight_Status'] == 'Delayed']['Airline'].value_counts()

print("Airline with the Most Delayed Flights:")
print(delayed_counts.idxmax())
print("Number of Delayed Flights:", delayed_counts.max())

Airline with the Most Delayed Flights:
Vistara
Number of Delayed Flights: 13


In [25]:
cancelled_counts = df[df['Flight_Status'] == 'Cancelled']['Airline'].value_counts()

print("Airline with the Most Cancelled Flights:")
print(cancelled_counts.idxmax())
print("Number of Cancelled Flights:", cancelled_counts.max())

Airline with the Most Cancelled Flights:
SpiceJet
Number of Cancelled Flights: 3


In [26]:
print("Total Number of Flights:", len(df))

print("\nFlights by Airline:")
print(df['Airline'].value_counts())

print("\nFlight Status Distribution:")
print(df['Flight_Status'].value_counts())

print("\nDelayed Flights by Airline:")
print(df[df['Flight_Status'] == 'Delayed']['Airline'].value_counts())

print("\nCancelled Flights by Airline:")
print(df[df['Flight_Status'] == 'Cancelled']['Airline'].value_counts())

Total Number of Flights: 180

Flights by Airline:
Airline
Vistara              37
SpiceJet             32
IndiGo               32
Akasa Air            31
Air India Express    24
Air India            24
Name: count, dtype: int64

Flight Status Distribution:
Flight_Status
On Time      114
Delayed       59
Cancelled      7
Name: count, dtype: int64

Delayed Flights by Airline:
Airline
Vistara              13
SpiceJet             12
IndiGo               11
Air India             9
Akasa Air             8
Air India Express     6
Name: count, dtype: int64

Cancelled Flights by Airline:
Airline
SpiceJet     3
Vistara      2
Air India    1
Akasa Air    1
Name: count, dtype: int64


## Key Findings and Observations

1. The dataset contains a total of **180 flight records**.

2. **Vistara** is the airline with the highest number of flights, with **37 flights**.

3. **On Time** is the most common flight status, with **114 flights**.

4. **Vistara** has the highest number of delayed flights, with **13 delayed flights**.

5. **SpiceJet** has the highest number of cancelled flights, with **3 cancelled flights**.

6. Out of 180 flights, **59 flights were delayed** and **7 flights were cancelled**, while **114 flights operated on time**.

7. The analysis shows that **Vistara, SpiceJet, and IndiGo** have relatively higher numbers of flight records compared with the other airlines in the dataset.

8. Grouping and filtering the data using Pandas helped identify patterns in **flight operations, delays, and cancellations across different airlines**.
